hooks/templatedatavalidator.js

```diff
@@ -0,0 +1,115 @@
+ * @returns {boolean} false if validator should throw an error
+ */
+module.exports = function (pageData) {
+  const jamboConfig = parseJamboConfig();
+  return isGlobalConfigValid(pageData.global_config) && isPageVerticalConfigValid(pageData, jamboConfig);
+}
+
+/**
+ * Validates global config for the page template
+ * 
+ * @param {Object} global_config 
+ * @returns {boolean}
+ */
+function isGlobalConfigValid(global_config) {
+  if (!global_config.experienceKey) {
+    error('Missing Info: no experienceKey found in config file.');
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Valivates vertical config for current page template
+ * 
+ * @param {Object} pageData 
+ * @param {Object} jamboConfig 
+ * @returns {boolean}
+ */
+function isPageVerticalConfigValid(pageData, jamboConfig) {
+  let isValid = true;
+  const themeDirectory = path.resolve(process.cwd(), jamboConfig.dirs.themes, jamboConfig.defaultTheme);
+  Object.keys(pageData.verticalsToConfig).forEach(key => {
+    if (key === 'Universal') {
+      console.log(pageData);
+      if (!isAllVerticalConfigsValid(pageData.verticalConfigs, jamboConfig)) {
+        isValid = false;
+      }
+      return;
+    }
+    const universalSectionTemplate = pageData.verticalsToConfig[key].universalSectionTemplate;
+    const cardType = pageData.verticalsToConfig[key].cardType;
+    if (!isUniversalSectionTemplateValid(key, themeDirectory, universalSectionTemplate)
+      | !isCardTypeValid(key, themeDirectory, cardType)) {
+      isValid = false;
+    }
+  });
+  return isValid;
+}
+
+/**
+ * If universalsectiontemplate is defined, check whether the corresponding file exists in theme
+ * 
+ * @param {string} verticalName
+ * @param {string} themeDir
+ * @param {string} template
+ * @returns {boolean}
+ */
+function isUniversalSectionTemplateValid(verticalName, themeDir, template) {
+  if (template) {
+    const universalSectionPath = path.resolve(themeDir, 'universalsectiontemplates/', template + '.hbs');
+    if (!fs.existsSync(universalSectionPath)) {
+      error(`Invalid universalSectionTemplate: can't find "${template}" at the expected path "${universalSectionPath}" for vertical "${verticalName}".`);
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * If cardType is defined, check whether the corresponding file exists in theme or custom card folder
+ * 
+ * @param {string} verticalName
+ * @param {string} themeDir
+ * @param {string} cardType
+ * @returns {boolean}
+ */
+function isCardTypeValid(verticalName, themeDir, cardType) {
+  if (cardType) {
+    const cardTypePath = path.resolve(themeDir, 'cards/', cardType);
+    const customCardTypePath = path.resolve('cards/', cardType);
+    if (!fs.existsSync(cardTypePath) && !fs.existsSync(customCardTypePath)) {
+      error(`Invalid cardType: can't find "${cardType}" at at the expected paths "${cardTypePath}" or "${customCardTypePath}" for vertical "${verticalName}".`);
```

Comment:
just curious but what does cardTypePath print out in sitesgitstorm console? maybe we would want just the relative path from the repo root with path.join instead of the absolute path

Reply:
it prints out `/repo-vol/themes/answers-hitchhiker-theme/cards/link-standard`

Reply:
ok I'm fine either way! I think just having it be `themes/answers-hitchhiker-theme/cards/link-standard` is better most of the time, but very occasionally it might be useful to have the full path

Reply:
hmm `path.join(hemeDir, 'cards/', cardType);` would still give me full path though

Reply:
oh I guess you would have to change themeDir to also use path.join()

Reply:
ah duh, yup updated!